infrastructures/openshift/src/test/java/org/eclipse/che/workspace/infrastructure/openshift/project/OpenShiftProjectFactoryTest.java

```diff
@@ -33,13 +61,211 @@
  */
 @Listeners(MockitoTestNGListener.class)
 public class OpenShiftProjectFactoryTest {
+
+  @Mock private OpenShiftClientConfigFactory configFactory;
   @Mock private OpenShiftClientFactory clientFactory;
+
+  @Mock
+  private NonNamespaceOperation<
+          Project, ProjectList, DoneableProject, Resource<Project, DoneableProject>>
+      projectOperation;
+
+  @Mock private OpenShiftClient osClient;
+
   private OpenShiftProjectFactory projectFactory;
 
+  @BeforeMethod
+  public void setUp() throws Exception {
+    lenient().when(clientFactory.createOC()).thenReturn(osClient);
+    lenient().when(osClient.projects()).thenReturn(projectOperation);
+  }
+
+  @Test(
+      expectedExceptions = ConfigurationException.class,
+      expectedExceptionsMessageRegExp =
+          "che.infra.kubernetes.namespace.default or "
+              + "che.infra.kubernetes.namespace.allow_user_defined must be configured")
+  public void
+      shouldThrowExceptionIfNoDefaultNamespaceIsConfiguredAndUserDefinedNamespacesAreNotAllowed()
+          throws Exception {
+    projectFactory =
+        new OpenShiftProjectFactory(
+            "projectName", "", "", null, false, clientFactory, configFactory);
+  }
+
+  @Test
+  public void shouldReturnDefaultProjectOnlyWhenItExistsAndUserDefinedIsNotAllowed()
+      throws Exception {
+    prepareNamespaceToBeFoundByName(
+        "che-default",
+        new ProjectBuilder()
+            .withNewMetadata()
+            .withName("che-default")
+            .withAnnotations(
+                ImmutableMap.of(
+                    PROJECT_DISPLAY_NAME_ANNOTATION,
+                    "Default Che Project",
+                    PROJECT_DESCRIPTION_ANNOTATION,
+                    "some description"))
+            .endMetadata()
+            .withNewStatus("Active")
+            .build());
+
+    projectFactory =
+        new OpenShiftProjectFactory(
+            "predefined", "", "", "che-default", false, clientFactory, configFactory);
+
+    List<KubernetesNamespaceMeta> availableNamespaces = projectFactory.list();
+    assertEquals(availableNamespaces.size(), 1);
+    KubernetesNamespaceMeta defaultNamespace = availableNamespaces.get(0);
+    assertEquals(defaultNamespace.getName(), "che-default");
+    assertEquals(defaultNamespace.getAttributes().get(DEFAULT_ATTRIBUTE), "true");
+    assertEquals(
+        defaultNamespace.getAttributes().get(PROJECT_DISPLAY_NAME_ATTRIBUTE),
+        "Default Che Project");
+    assertEquals(
+        defaultNamespace.getAttributes().get(PROJECT_DESCRIPTION_ATTRIBUTE), "some description");
+    assertEquals(defaultNamespace.getAttributes().get(PHASE_ATTRIBUTE), "Active");
+  }
+
+  @Test
+  public void shouldReturnDefaultProjectOnlyWhenItDoesNotExistAndUserDefinedIsNotAllowed()
```

Comment:
Again, the names of `shouldReturnDefaultProjectOnlyWhenItExistsAndUserDefinedIsNotAllowed` and `shouldReturnDefaultProjectOnlyWhenItDoesNotExistAndUserDefinedIsNotAllowed` seem to contradict each other.

Reply:
`only when` replaced with `when`